pkg/adaptation/default-validator.go

```diff
@@ -0,0 +1,173 @@
+		log.Infof(noCtx, "built-in NRI default validator is disabled")
+		return nil
+	}
+
+	v := NewDefaultValidator(cfg)
+	return &BuiltinPlugin{
+		Base:  "default-validator",
+		Index: "00",
+		Handlers: BuiltinHandlers{
+			ValidateContainerAdjustment: v.ValidateContainerAdjustment,
+		},
+	}
+}
+
+// NewDefaultValidator creates a new instance of the default validator plugin.
+func NewDefaultValidator(cfg *DefaultValidatorConfig) *DefaultValidator {
+	return &DefaultValidator{cfg: *cfg}
+}
+
+func (v *DefaultValidator) ValidateContainerAdjustment(ctx context.Context, req *ValidateContainerAdjustmentRequest) (*ValidateContainerAdjustmentResponse, error) {
+	if err := v.validateAdjustment(ctx, req); err != nil {
+		log.Errorf(ctx, "rejecting adjusted container: %v", err)
+		return &ValidateContainerAdjustmentResponse{
+			Reject: true,
+			Reason: err.Error(),
+		}, nil
+	}
+
+	if err := v.validatePlugins(req); err != nil {
+		log.Errorf(ctx, "rejecting adjusted container: %v", err)
+		return &ValidateContainerAdjustmentResponse{
+			Reject: true,
+			Reason: err.Error(),
+		}, nil
+	}
+
+	return &ValidateContainerAdjustmentResponse{}, nil
+}
+
+func (v *DefaultValidator) validateAdjustment(_ context.Context, req *ValidateContainerAdjustmentRequest) error {
+	if !v.cfg.RejectOCIHooks || req.Adjust == nil {
+		return nil
+	}
+
+	if err := v.validateOciHooks(req); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v *DefaultValidator) validateOciHooks(req *ValidateContainerAdjustmentRequest) error {
+	if req.Adjust == nil {
+		return nil
+	}
+
+	if plugins, claimed := req.Owners.HooksOwner(req.Container.Id); claimed {
+		what := "plugin"
+		if strings.Contains(plugins, ",") {
+			what = "plugins"
+		}
+
+		return fmt.Errorf("%w: %s %s attempted restricted hook injection",
+			ErrValidation, what, plugins)
+	}
+
+	return nil
+}
+
+func (v *DefaultValidator) validatePlugins(req *ValidateContainerAdjustmentRequest) error {
+	var (
+		annotations = req.GetPod().GetAnnotations()
+		container   = req.GetContainer().GetName()
+		required    []string
+	)
+
+	value, ok := getEffectiveAnnotation(annotations, RequiredPluginsAnnotation, container)
+	if !ok {
+		return nil
+	}
+
+	if err := yaml.Unmarshal([]byte(value), &required); err != nil {
+		return fmt.Errorf("invalid %s annotation %q: %w", RequiredPluginsAnnotation, value, err)
+	}
+
+	if len(required) == 0 {
+		return nil
+	}
+
+	present := map[string]struct{}{}
+	for _, p := range req.Plugins {
+		if p != nil {
+			present[p.Name] = struct{}{}
+		}
+	}
+
+	for _, r := range required {
+		if _, ok := present[r]; !ok {
+			return fmt.Errorf("%w: required plugin %q not present", ErrValidation, r)
+		}
+	}
+
+	return nil
+}
+
+func getEffectiveAnnotation(annotations map[string]string, key, container string) (string, bool) {
+	if len(annotations) == 0 {
+		return "", false
+	}
+
+	keys := []string{
+		key + "/container." + container,
+		key + ".pod",
+		key,
+	}
+
+	for _, key := range keys {
```

Comment:
This is masking the `key` argument above. Could we rename one or the other to clarify the difference?

Reply:
Updated to use `k` and `v` instead of `key` and `value`.